Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,230 +1,244 @@
         if (testKey != expectedKey || testValue != expectedValue) {
             return false;
         }
     }
     return true;
 }
 
-int hashing(int key) {
-    return key % SIZE;
+// hashing 
+// returns a "hashed" index based on modulus of the key's ASCII integer sum and the SIZE of the hashTable
+int hashing(HashTable *hashtable, char *key ) {
+    int i = 0;
+    int count = 0;
+    while (key[i] != NULL) {
+        count += key[i];
+        i++;
+    }
+    return count % hashtable->size;
 }
+
+
 // creates a new Item struct
 // returns new struct
-Item * createItem(int key, int value, int success) {
+Item * createItem(char key[], int value, int * success) {
     Item * item = malloc(sizeof(Item));
-    if (item == NULL) {
-    } else {
-        item->key = key;
+    if (item != NULL) {        
+        item->key = strdup(key);
         item->value = value;
-        success = SUCCESS;
+        *success = SUCCESS;
         return item;
     }
+    return NULL;
 }
 
-// insert takes in a key and value
-// calls a helper function to create an Item struct with the provided key and value
-// adds the new Item to the hashTable
-// if the expected value is in the hashTable at the expected key then returns a success code
-void insert(int type, int key, int value, int success) {
-    if (type == TEST) {
-        testHash[key] = createItem(key, value, success);
-        if (testHash[key]->value != value) {
-            success = ERROR;
-        }
-        
-    } else if (type == EXPECTED) {
-        expectedHash[key] = createItem(key, value, success);
-        if (expectedHash[key]->value != value) {
-            success = ERROR;
-        }
+// creates a new HashTable struct
+// returns new HashTable struct
+HashTable * createHashTable(int size) {
+    if (size < 1 || size > 100) return NULL;
+    HashTable * hashTable = malloc(sizeof(HashTable));
+    hashTable->items = malloc(sizeof(Item) * size);
+    if (hashTable == NULL || hashTable->items == NULL) {
+        return NULL;
     }
+    hashTable->size = size;
+    for (int i = 0; i < size; i++) {
+        hashTable->items[i] = NULL;
+    }
+    return hashTable;
 }
 
-int testInsert() {
-    return ERROR;
-}
 
-void buildInsertTests() {
-    //InsertTests
-    // 5 passes
-    // 5 Fails
-    // Empty Array
-    // Too large for hash table
-    // Collision
-    // Full Hash Table
+// insert 
+// adds the new Item struct to the hashTable
+// if the expected value is in the hashTable at the expected key then returns a success code
+int insert(HashTable * hashTable, char key[], int value) {
+    int success = ERROR;
+    int index = hashing(hashTable, key);
+    if (isCollision(hashTable, index)) {
+        //find next free space  
+        while(hashTable->items[index] != NULL && hashTable->items[index]->key != key && hashTable->items[index]->key != -1) 
+        { 
+            index++; 
+            index % hashTable->size; 
+        } 
```

Comment:
If size + 1 elements are inserted, what is the current behavior of this collision mitigation code?

Reply:
Updated to check to see if index was <= testHash->size before preceeding with index++. Will expand on this limit when building Insert testing.
[commit aafcf27645ea4097ae6559808d5307e70f99ae1a](https://github.com/kcarrel/Training-Notes/pull/12/commits/aafcf27645ea4097ae6559808d5307e70f99ae1a)